src/handlers/psp.ts

```diff
@@ -0,0 +1,436 @@
+  if (isEmpty(evmChains)) {
+    throw new Error('One or more evm compatible chain(s) must be defined in the provided config');
+  }
+  const providerPromises = flatMap(
+    evmChains.map((chain: ChainConfig) =>
+      map(chain.providers, async (chainProvider, providerName) => {
+        const providerLogOptions: node.LogOptions = {
+          ...baseLogOptions,
+          meta: {
+            ...baseLogOptions.meta,
+            chainId: chain.id,
+            providerName,
+          },
+        };
+
+        // Initialize provider specific data
+        const [logs, evmProviderState] = await initializeProvider(chain, chainProvider.url || '');
+        node.logger.logPending(logs, providerLogOptions);
+        if (isNil(evmProviderState)) {
+          node.logger.warn('Failed to initialize provider', providerLogOptions);
+          return null;
+        }
+
+        return {
+          airnodeWallet,
+          chainId: chain.id,
+          providerName,
+          ...evmProviderState,
+        } as ProviderState<EVMProviderState>;
+      })
+    )
+  );
+
+  const providerStates = await Promise.all(providerPromises);
+  const validProviderStates = providerStates.filter((ps) => !isNil(ps)) as ProviderState<EVMProviderState>[];
+
+  return { ...state, providerStates: validProviderStates };
+};
+
+const checkSubscriptionsConditions = async (
+  subscriptions: Id<Subscription>[],
+  apiValuesBySubscriptionId: { [subscriptionId: string]: ethers.BigNumber },
+  contract: ethers.Contract,
+  voidSigner: ethers.VoidSigner,
+  logOptions: node.LogOptions
+) => {
+  const validSubscriptions: CheckedSubscription[] = [];
+  const conditionPromises = subscriptions.map(
+    (subscription) =>
+      checkSubscriptionCondition(subscription, apiValuesBySubscriptionId[subscription.id], contract, voidSigner).then(
+        ([logs, isValid]) => [logs, { subscription, isValid }]
+      ) as Promise<node.LogsData<{ subscription: Id<Subscription>; isValid: boolean }>>
+  );
+  const result = await Promise.all(conditionPromises);
+  result.forEach(([log, data]) => {
+    const subscriptionLogOptions: node.LogOptions = {
+      ...logOptions,
+      additional: {
+        ...logOptions.additional,
+        subscriptionId: data.subscription.id,
+      },
+    };
+    node.logger.logPending(log, subscriptionLogOptions);
+    if (data.isValid) {
+      validSubscriptions.push({
+        ...data.subscription,
+        apiValue: apiValuesBySubscriptionId[data.subscription.id],
+      });
+    }
+  });
+
+  return validSubscriptions;
+};
+
+const groupSubscriptionsBySponsorWallet = async (
+  subscriptionsBySponsor: Dictionary<CheckedSubscription[]>,
+  config: Config,
+  provider: ethers.providers.Provider,
+  currentBlock: number,
+  providerLogOptions: node.LogOptions
+): Promise<SponsorWalletWithSubscriptions[]> => {
+  const sponsorWalletsWithSubscriptions: SponsorWalletWithSubscriptions[] = [];
```

Comment:
I can't help myself:
```
  const sponsorAddresses = Object.keys(subscriptionsBySponsor);
  const sponsorWalletAndTransactionCountPromises = sponsorAddresses.map(
    (sponsor) =>
      getSponsorWalletAndTransactionCount(config, provider, currentBlock, sponsor).then(([logs, data]) => [
        logs,
        {...data, sponsor},
      ]) as Promise<node.LogsData<(SponsorWalletTransactionCount | null) & { sponsor: string }>>
  );
  const sponsorWalletsAndTransactionCounts = await Promise.all(sponsorWalletAndTransactionCountPromises);

  const makeSponsorLogOptions = (extra: {}) => ({
    ...providerLogOptions,
    additional: {
      ...providerLogOptions.additional,
      ...extra
    }
  });

  return sponsorWalletsAndTransactionCounts.filter(([_logs, data]) => {
    if (isNil(data.sponsorWallet) || isNil(data.transactionCount)) {
      node.logger.warn('Failed to fetch sponsor wallet or transaction count', makeSponsorLogOptions({sponsor: data.sponsor}));
      return;
    }

    return true;
  })
    .map(([logs, data]) => {
      node.logger.logPending(logs, makeSponsorLogOptions({sponsor: data.sponsor}));

      return {
        subscriptions: subscriptionsBySponsor[data.sponsor].map((subscription) => ({
          ...subscription,
          nonce: data.transactionCount,
        })),
        sponsorWallet: data.sponsorWallet,
      };
    });
};
```

Reply:
Those log options are horrible... perhaps you could have a set of factory(?) functions that allow you to compose them more easily with less duplication (like above).

Reply:
Indeed they are 😆  and they add a bunch of lines to the code but the end result is really nice:

![image](https://user-images.githubusercontent.com/5776880/157868357-501b7f3d-5dad-42d2-98c2-055dd36d580b.png)

I'm going to think about how to improve it. Open to suggestions

Reply:
Sorry, my comment may have come across severely.
I support you using the existing logging solution, in fact I think it's very cool that you did, but I suggest you use a function to help you compose the logOptions so that there's less duplication. I see there's an open Jira task to make it so that the log subsystem is configured up-front instead of with every call.